Fixed choropleth

FloChehab opened the pull request:
Closes #971, closes #996 and closes #905.

Other changes:

- I removed the chunk below, as I think this check adds an unwanted constraint (if too many colors are _asked_ then it will fail in `branca.utilities.color_brewer`):

```python
        if threshold_scale is not None and len(threshold_scale) > 6:	
            raise ValueError('The length of threshold_scale is {}, but it may '	
                             'not be longer than 6.'.format(len(threshold_scale)))  # noqa
```

- I have updated the docstring.
- I have made few changes linked to the fact that every time `color_brewer` was used, one extra color was generated.


One thing I am not sure: I have added the `nan_fill_color` parameter at a _logical_ place in the list of parameters but we might want to change that.


All current tests should pass. But new tests might be welcomed for the new _features_.

FloChehab commented on folium/folium.py:
```diff
@@ -550,32 +553,39 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
                         get_by_key(obj.get(key.split('.')[0], None),
                                    '.'.join(key.split('.')[1:])))
 
             def color_scale_fun(x):
```
Here, I decided to mimic the previous behavior described in the comment below in order not to introduce any braking change:

```python
# we consider that values outside the color 
# should be affected to the first (or last) color bin.
```

However, doing so is actually producing misleading maps (for example, if we force the color scale do go from 1 to 10 with step 1 and we have -60 as a value, it would map to the same color as the bin 1 to 2 without any warning).
It might be a good idea to raise a warning in such circumstances or just to forbid this behavior (that's what I would do).
I didn't choose the last option as it was causing unit tests to fails:

Some notebooks use `branca` `split_six` functions which can produce *weird* intervals, for example:
```python
import pandas as pd
from folium import branca
threshold_scale = branca.utilities.split_six(series=pd.Series([1,2,2,3,3,3,3,3,3,4]))
# produces (as expected): [1.0, 3.0, 3.0, 3.0, 3.0]
```
Therefore, the `4` we would like to map to a color would be outside of the `threshold_scale`.

Conengmo replied:
Good that you point this out. We have to make a decision here. I'm listing the options:

1. Let values outside of the range use the min or max value (current behavior).
  * This makes sense if you have a few outliers you don't want to stretch out your range for.
  * No time to filter out outliers, get stuff done quickly.
2. Raise a UserWarning when values fall outside of the range.
  * Sort of compromise between the current behavior and raising an error.
  * Numpy does this a lot, which results in people figuring out how to disable/ignore them.
  * Kind of annoying (subjective, I know :p )
3. Raise an exception when values fall outside of the range.
  * Having a value lie outside of the range is a mistake and we don't do mistakes.
  * Force the user to give either a good range or remove outliers.
4. Give the values outside the range the same color as `nan_fill_color`.
  * This does show they are not in the range, but it makes them indistinguishable from nan values.
5. Give the values outside the range a special color and add a new argument for it.
  * Makes it easy to see which values fall outside the range.
  * Adds complexity, another argument to check out and set.

I'm currently leaning towards raising an exception. Many users may not pass a custom range, so they will not encounter this issue. If a user passes a custom range, that means he or she is advanced enough to also be able to properly preprocess the data and deal with outliers.

On the other hand I like how folium is hackable. If someone wants to deal with their outliers by cheating a bit, who are we to judge. But then maybe a UserWarning would be appropriate.

So after writing this comment I think I would lean towards leaving it as is and add a UserWarning. Do you want to make the case for another option? I'd like to hear your opinion. We can change the tests based on what we decide.

Conengmo replied:
Let's raise a `ValueError` then, with a useful message. We'll learn our users some proper preprocessing ;) Better to be explicit than have it fail silently.

Conengmo commented on folium/folium.py:
```diff
@@ -466,14 +469,17 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
             start with 'feature' and be in JavaScript objection notation.
             Ex: 'feature.id' or 'feature.properties.statename'.
         threshold_scale: list, default None
```
Not your change, but I would like that all arguments that have a default value of `None` because they are optional to have a docstring like `name: type, optional` instead of `name: type, default None`. In this case the default is actually a list with 7 values.

FloChehab replied:
I didn't get what you are saying here, could you clarify a bit? Thanks

Conengmo replied:
In many cases the default value is None, but from the users perspective that's not really helpful since None disables the argument. In that case it's better to say the argument is optional.

But here that's not really the case since the default value should actually be a list of 7 values so I guess my point doesn't really work here.

This doesn't matter here so I'll set the conversation as resolved :)

Conengmo commented on folium/folium.py:
```diff
@@ -550,32 +553,39 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
-                )
-                return color_range[idx-1]
+                key_of_x = get_by_key(x, key_on)
+                if key_of_x in color_data.keys():
```
I was thinking if this could be made a bit simpler. Instead of two if-else statements, make two if-statements that return on something wrong and keep the happy path unindented. For example here return immediately if the value is missing, no comment needed.
```suggestion
                if key_of_x not in color_data.keys():
                    return nan_fill_color
```

Then set the value, do a if-statement whether to return on `np.isnan`, and do the rest.

This is fine as it is, so no problem if you leave it like this. I like to discuss the finishing touch, that's all.

FloChehab replied:
I'll apply the change in the next commit.

FloChehab replied:
Should be good now

Conengmo replied:
Looks good indeed, thanks.

FloChehab commented on folium/folium.py:
```diff
@@ -466,14 +469,17 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
             start with 'feature' and be in JavaScript objection notation.
             Ex: 'feature.id' or 'feature.properties.statename'.
         threshold_scale: list, default None
```
I was suggesting `bins` following my experience with Pandas [`hist` function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.hist.html) (similar to Matplotlib).
We could have it set default to `6` and make it accept a number (of bins) or a list of numbers considered as the edges of the bins (just like Pandas).

So if I understand you correctly, we could add a `bins` parameter, and if it is not set would default to `threshold_scale`?

I'll wait for your "go/no go" on this part. (I've noted the need to remove `Data range for D3 threshold scale.`)

Conengmo replied:
Nice idea to make it behave as in Pandas hist. It's the same as the Numpy hist bins argument isn't it?
https://docs.scipy.org/doc/numpy/reference/generated/numpy.histogram.html

The way to go here IMO is to add `bins` as an argument, removing `threshold_scale` while keeping backwards compatibility. So `threshold_scale` shouldn't be listed in the documentation but still work. One way to do this is to accept `**kwargs` and retrieve `threshold_scale` from it. For example:

```
def choropleth(bins=6, **kwargs):
    if 'threshold_scale' in kwargs:
        bins = threshold_scale
        raise DeprecationWarning('maybe?')
```

Is this what you had in mind?

FloChehab replied:
> Is this what you had in mind?
Seems better than what I had in mind! (I am not a Python master yet :smile: )

> It's the same as the Numpy hist bins argument isn't it?
I think it is the same spirit.

Conengmo replied:
Oops example should be:
`bins = kwargs['threshold_scale']`

Conengmo commented:
By the way, have you considered working with opacity instead of coloring for missing/nan/out-of-range values?

Conengmo commented on folium/folium.py:
```diff
@@ -614,9 +624,10 @@ def highlight_function(x):
-                brewed[1:len(color_domain)],
+                brewed,
                 index=color_domain,
                 vmin=color_domain[0],
```
Are you sure? It seems they are passed to the init of the parent class, `ColorMap`. That one uses them in its `render` function.

FloChehab replied:
Indeed, I was wrong, I reverted that part.

FloChehab commented:
> By the way, have you considered working with opacity instead of coloring for missing/nan/out-of-range values?

I think those values should standout that's why I went for `'black'`, but if someone wants to play with opacity it should be possible with HEX strings `nan_fill_color='#ffffff00'` (it works, I just tested). We could mention it in the docstring.

FloChehab commented:
I have updated the code accordingly to our discussions, I'll wait for your feedback.

Obviously, CI will fail because of the `DeprecationWarning` and the new check to see if all values fall into the domain defined by bins. So next we will have on how to update the current tests. Then, what new tests to add. And finally, as you said, give new examples.

FloChehab commented on folium/folium.py:
```diff
@@ -508,24 +522,28 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
+            raise DeprecationWarning('''
+                choropleth `threshold_scale` parameter is now depreciated
+                in favor of the `bins` parameter.''')
+
```
I am not sure that:
```python
            raise DeprecationWarning('''
                choropleth `threshold_scale` parameter is now depreciated
                in favor of the `bins` parameter.''')
```

Is the ideal way to do it as it will stop the script as an error.

I have looked online and tried:
```python
warnings.warn('''
                choropleth `threshold_scale` parameter is now depreciated
                in favor of the `bins` parameter.''', DeprecationWarning)
```

But it doesn't seem to be shown in the Notebook I am using (it works well in the unit test on the other hand).

I am sure you have a suggestion :)

Conengmo replied:
You're right that the warning shouldn't be raised, instead `warnings.warn(msg, category)` is the way to go. Good find.

Deprecation warnings are ignored by default. They only appear in systems that enable them. We could use a `FutureWarning`, but I think I'm more inclined to use a `DeprecationWarning` and then once we later really remove the feature raise an error.
https://docs.python.org/3/library/warnings.html#warning-categories

FloChehab replied:
I have updated the code, we should be good

FloChehab commented:
By the way, while we are working on choropleth, we could also allow the `fill_color` parameter to be a direct mapping of colors for each bin defined. This would enable the creation of more custom maps (eg: with two colors).

Conengmo commented on folium/folium.py:
```diff
@@ -465,15 +467,27 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
+            including the rightmost edge, allowing for non-uniform bin widths.
+            For more information on this parameter, please have a look at
+            numpy.histogram function.
+            Warning: if any (non-nan) values is outside of the inclusive
```
IMO you don't really have to state that here, they will see the error when it happens. Makes the docstring a bit less lenghty also.

FloChehab replied:
Better ?

Conengmo commented on folium/folium.py:
```diff
@@ -538,55 +556,73 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
-        else:
-            color_domain = None
+        if color_data is not None and key_on is not None:
+            # numpy histogram_bin_edges is only available in the most
```
No need for this comment

Conengmo commented on folium/folium.py:
```diff
@@ -538,55 +556,73 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
+            # recent numpy version, so we will use numpy histogram
+            real_values = np.array(list(color_data.values()))
+            real_values = real_values[~np.isnan(real_values)]
+            _, bins = np.histogram(real_values, bins=bins)
```
I'm not a fan of reusing variable names when their type may have changed. Can we use a different name? In Numpy they name the output `bin_edges`.

FloChehab replied:
I went for `bin_edges`

Conengmo commented on folium/folium.py:
```diff
@@ -538,55 +556,73 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
+
+            bins_min, bins_max = min(bins), max(bins)
+            if np.any((real_values < bins_min) | (real_values > bins_max)):
+                raise ValueError('''
```
Because of the triple quotes the error includes new lines. Better to break up the string over the lines something like this:

```
raise ValueError(
    'All values are expected to fall into one of the provided '
    'bins (or to be Nan). Please check the `bins` parameter '
    'and/or your data.'
)
```

FloChehab replied:
I have made the change. I wasn't aware of this way to split strings in python, thanks !

Conengmo commented on folium/folium.py:
```diff
@@ -538,55 +556,73 @@ def choropleth(self, geo_data, data=None, columns=None, key_on=None,
+
+                if key_of_x not in color_data.keys():
+                    # 1: opacity is handled within nan_fill_color
+                    return nan_fill_color, 1
```
Having `nan_fill_color` as RGBA values adds some complexity. What if someone uses a regular RGB value and expects the `fill_opacity` to work on these fields as well? Do we have to check if `nan_fill_color` is a RGBA value and only return `opacity = 1` in that case?

I'm skeptical of adding more complexity by parsing color strings. Either return `opacity 1` or go back to the RGB situation we had before. What do you think?

Conengmo replied:
Or maybe instead add a `nan_fill_opacity` argument? That would be more in line with the rest of the code. We don't use RGBA in any other place I think.

FloChehab replied:
I think, the easiest way to go would be to add a `nan_fill_opacity` parameter (that would default to the `fill_opacity` value if it's not set) to the function because I am sure at some point in the future someone is going to ask for such possibility.
And going that way we won't have to think about tricky HEX RGBA strings.

What do you think ?

FloChehab replied:
Oh we thought at the same thing... I'll add the parameter.

Conengmo replied:
I agree, that would be best. 

I do like the idea of using RGBA though, so thanks for introducing that. But we would have to handle that more consistently then. Maybe for in another PR.

FloChehab replied:
I have updated the code to have a `nan_fill_opacity` parameter.

Conengmo commented:
> we could also allow the fill_color parameter to be a direct mapping of colors for each bin defined

That's not a bad idea. Is that something you would want to do yourself? In any case, I suggest to do that in a new PR after this one, to make this one manageable.

Conengmo commented:
Neat! Nice work. I think this is a well rounded PR, with fixing the binning and missing values.

So, what's next? Updating the examples I would say. Only the notebook tests failed, probably because we deprecated `threshold_scale`.

I know of two notebooks that have choropleth examples, but there may be others.

https://nbviewer.jupyter.org/github/python-visualization/folium/blob/master/examples/GeoJSON_and_choropleth.ipynb

https://nbviewer.jupyter.org/github/python-visualization/folium/blob/master/examples/Quickstart.ipynb

Another thing would be to look at the tests and see if we can change/add something for the new and updated functionality.

FloChehab commented:
I have updated the notebooks and added one test (I did my best to come up with something straightforward for testing the new functionalities, it might not be up to your standards).

Also, I am wondering why are the notebooks stored on Github with runned cells? (I feel that it is quite a waste of space, and they are not all up-to-date)

Conengmo commented:
Link to nbviewer for your notebook changes:
http://nbviewer.jupyter.org/github/FloChehab/folium/blob/fix_choropleth/examples/Quickstart.ipynb
http://nbviewer.jupyter.org/github/FloChehab/folium/blob/fix_choropleth/examples/GeoJSON_and_choropleth.ipynb

> I have updated the notebooks and added one test

Good! Notebooks look good and your test seems decent. Thanks. I think we're done with this PR then! What do you think, good to go?

> I am wondering why are the notebooks stored on Github with runned cells

Good question. It's so they are immediately viewable with something like nbviewer.jupyter.org. I don't really like the diffs being unreadable. We do let Travis test if they are renderable though.

Do you have an idea how the way we deal with the notebooks could be improved?

ocefpaf commented:
> I would consider the PR as done on my side too.
> 
> I hadn't taken a look at `nbviewer` before and therefore, I get why notebooks are stored with run cells in the repo.
> 
> I am more used to Gitlab CI, and in such context I would convert the notebooks to html in the pipeline and store them as job artefacts (which can be directly accessible using a simple URL). I don't know if this could be possible on Github + Travis.

It is and I used to do that but nbviewer is quite handy, it reduces the maintenance  steps and we don't waste CIs running the notebooks.

ocefpaf commented:
BTW, locally I use `nbdime` to check notebooks diffs on git. It is quite handy to review notebooks commits.

Conengmo commented:
Merged! Thanks @FloChehab, I've enjoyed working with you.

FloChehab commented:
@Conengmo you are welcome ! It was a pleasure working with you too (you took quite some time discussing, commenting and giving advice !)

It might be my only contribution to Folium: we had to use Folium for a course at university when we faced the now fixed bugs related to `choropleth`; I came up with some nasty fix and wanted to have a more durable fix for the whole community :smile: Anyway, it was very nice to collaborate !

ReidWeb commented:
Has a version with this been published yet?

ocefpaf commented:
> Has a version with this changeset been published yet? Or is there a way to easily install the pre-release version? Would quite like to utilise the updated version for something.

Not yet. Planning on a new release for next week.
